### PR TITLE
RavenDB-21142 TreeCursor: allocate FastStack via ObjectPool to avoid expensive Array.Resize calls

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21142

### Type of change

- Optimization

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Voron/Data/BTrees/TreeCursor.cs [hunk 1]
```diff
@@ -9,17 +9,32 @@ namespace Voron.Data.BTrees
 {
     public sealed class TreeCursor : IDisposable
     {
-        public FastStack<TreePage> Pages = new FastStack<TreePage>();
-
-        private static readonly ObjectPool<Dictionary<long, TreePage>> _pagesByNumPool = new ObjectPool<Dictionary<long, TreePage>>(() => new Dictionary<long, TreePage>(50, NumericEqualityComparer.BoxedInstanceInt64));
-
-        private readonly Dictionary<long, TreePage> _pagesByNum;
-
+        private class TreeCursorState
```
Comment (ppekrol) [inline src/Voron/Data/BTrees/TreeCursor.cs]: sealed?

**Review by ppekrol — APPROVED** [review]
LGTM, but any before/after numbers?

File: src/Voron/Data/BTrees/TreeCursor.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Data/BTrees/TreeCursor.cs]: yeah

File: src/Voron/Data/BTrees/TreeCursor.cs [hunk 2]
```diff
@@ -48,56 +63,56 @@ public void Update(FastStack<TreePage> stack, TreePage newVal)
...
+            var removedPrimary = _state.PageByNum.Remove(p.PageNumber);
             var removedSecondary = false;
 
             if (_anyOverrides)
             {
                 var pagesNumbersToRemove = new HashSet<long>();
 
-                foreach (var page in _pagesByNum.Where(page => page.Value.PageNumber == p.PageNumber))
+                foreach (var page in _state.PageByNum.Where(page => page.Value.PageNumber == p.PageNumber))
                     pagesNumbersToRemove.Add(page.Key);
```
Comment (redknightlois) [inline src/Voron/Data/BTrees/TreeCursor.cs:112]: Isn't `_state.PageByNum.Where(page => page.Value.PageNumber == p.PageNumber)` just `_state.PageByNum[p.PageNumber]` ?

Comment (ayende) [reply src/Voron/Data/BTrees/TreeCursor.cs:112]: No, that is why we may have overrides.

**Review by ayende — APPROVED** [review]

File: src/Voron/Data/BTrees/TreeCursor.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Voron/Data/BTrees/TreeCursor.cs]: done

**Review by arekpalinski — APPROVED** [review]

**Comment by maciejaszyk** [comment]
> LGTM, but any before/after numbers?

I compared a couple of indexing batches but I'm not sure if those records are valid since measuring this has a few challenges.
1) Batches should be equal so this would require writing own benchmark. (since we want to compare it 1:1)
2) Since it uses a cache it may vary on previously executed operations (for example enough space in cached FastStacks)

I compared a couple of them, seems we may gain "something"
image

We already had cache in `TreeCursor` for Dictionary so moving FastStack into it was very natural behaviour.

**Comment by ayende** [comment]
`-700 ms` in `TryGetNextValue` is pretty significant.

**Pull request merged**